sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py

```diff
@@ -0,0 +1,491 @@
+      elif mode == "overwrite":
+        sub_client.delete_subscription(subscription_path)
+        self.subscription = sub_client.create_subscription(
+            subscription_path, self.topic.name)
+      else:
+        self.subscription = sub_client.get_subscription(subscription_path)
+
+    snapshot_path = sub_client.snapshot_path(self.project, self.topic_name)
+    try:
+      self.snapshot = sub_client.create_snapshot(snapshot_path,
+                                                 self.subscription.name)
+    except gexc.AlreadyExists:
+      if mode == "error":
+        raise IOError("Snapshot '{}' already exists.".format(snapshot_path))
+      elif mode == "overwrite":
+        sub_client.delete_snapshot(snapshot_path)
+        self.snapshot = sub_client.create_snapshot(snapshot_path,
+                                                   self.subscription.name)
+      else:
+        # See: https://github.com/googleapis/google-cloud-python/issues/8554
+        self.snapshot = pubsub.types.Snapshot()
+        self.snapshot.name = snapshot_path
+    except gexc.MethodNotImplemented:
+      self.snapshot = None
+
+    self._child_subscriptions.append(self.subscription)
+    self._init_finalizers()
+
+  @property
+  def timestamp(self):
+    return self._timestamp
+
+  @property
+  def persist(self):
+    return self._persist
+
+  @persist.setter
+  def persist(self, persist):
+    self._persist = persist
+    self._init_finalizers()
+
+  def reader(self, seek_to_start=True, **kwargs):
+    self._assert_topic_exists()
+
+    reader_kwargs = self._reader_kwargs.copy()
+    reader_kwargs.update(kwargs)
+
+    if "subscription" not in reader_kwargs:
+      reader_kwargs["subscription"] = self._create_child_subscription(
+          seek_to_start=seek_to_start).name
+
+    reader = PatchedPubSubReader(self, self._reader_class, **reader_kwargs)
+    return reader
+
+  def writer(self):
+    self._timestamp = time.time()
+    writer = PatchedPubSubWriter(self, self._writer_class, (self.location,),
+                                 self._writer_kwargs)
+    return writer
+
+  @contextlib.contextmanager
+  def read_to_queue(self,
+                    seek_to_start=True,
+                    **kwargs):
+    self._assert_topic_exists()
+
+    reader_kwargs = self._reader_kwargs.copy()
+    reader_kwargs.update(kwargs)
+
+    if "subscription" in reader_kwargs:
+      created_subsciption = None
+    else:
+      created_subsciption = self._create_child_subscription(
+          seek_to_start=seek_to_start)
+      reader_kwargs["subscription"] = created_subsciption.name
+      assert created_subsciption in self._child_subscriptions
+
+    @functools.total_ordering
+    class PrioritizedTimestampedValue(TimestampedValue):
+
+      def __lt__(self, other):
+        return self.timestamp < other.timestamp
+
+    # Set arbitrary queue size limit to prevent OOM errors.
+    parsed_message_queue = queue.PriorityQueue(1000)
```

Comment:
Why a PriorityQueue? I don't think it can guarantee that messages yielded will have non-decreasing timestamp values.

Reply:
From my understanding, it does guarantee this. I had to implement the `PrioritizedTimestampedValue` class above so that values are yielded in non-decreasing order based on their timestamps rather than their values.